### PR TITLE
Roles 0.2.0/flush out the stream on wearer change

### PR DESCRIPTION
**Pull request by mudrila**
This PR introduces following changes/functionality:

- Fetch and Display Payroll and Vesting information for the given Hat Account
- Rinse out payroll/vesting stream to the current wearer when changing hat wearer
- Rinse out payroll/vesting stream to the current wearer and cancel stream, which automatically returns remaining asset balance to the Safe, when deleting hat

Testing:
1. If you were creating payroll before - make sure it appears in the Roles list.
2. Try changing the wearer on the hat with active payroll (aka start date and first tranche already passed). Expected result - funds that weren't claimed are transferred to the "current wearer" and then wearer is changed to the new wearer
3. Try deleting the hat with active payroll. Expected result - funds that weren't claimed are transferred to the "current wearer", remaining funds from the stream are transferred back to the Safe, stream is cancelled, and then hat is "deleted"

### CODE REVIEW

**Comment by tomstuart123** [comment]
@mudrila - great question. I don't think enough thought has been put into this. 

**Before I choose,**
1) I think regardless of the option, both options should have more user feedback during creation to make sure its clear.

I.e. if we want the user to input '1000 tokens' + '10 months', we should dynamically generate some copy saying 'That'll be 100 per month'. OR if we go the other path, if we want the user to input '100 tokens' over '10 months', we should generate some copy saying 'That'll remove 1000 tokens from the treasury'

2) I think the copy here and here both need clarifying to make it easier to understand
Screenshot 2024-07-22 at 14 44 21
Screenshot 2024-07-22 at 14 47 18


**Now my take on choice**

**re; Vesting** - this should 100% be the 'Total Amount'. Very few people think about vested tokens in a bottom up way 

**re; Payroll -** this could be either way... but I lean to the bottom up way due to i) I assume multiplication is easier than division (don't have to worry about decimals) and ii) when contractors quote their pay, they quote how much $x every y month. HOWEVER, this path relies on the dynamic text above (i.e. they could get confused if getting an 'invalid' message due to the total being bigger than what is in the treasury)

Note - @mudrila - if it's easier to make both Total Amount bottom down as a start, we can make this a seperate PR

@nicolaus-sherrill what's your take?

**Comment by tomstuart123** [comment]
@mudrila just testing this and the payroll I created last time isn't appearing [here](https://deploy-preview-2165.app.dev.decentdao.org/treasury?dao=sep:0x6922024B936f1e5485326044097d69C6DF87E8AF). Saying that, not sure if its because the payroll is no longer running ( I can't remember when I made it for)

Let me create a new role with payroll and let you know

**Comment by tomstuart123** [comment]
I haven't been able to test further than seeing the payroll I'm afraid (I'm getting the 'proposal failed' and treasury not updating to what's in production again... moralis?)

Anyway, just catching Adam up. So will again try after that. usually it resolves itself

**Review by DarksightKellar — APPROVED** [review]
Code review done, but not yet manually tested

**Comment by tomstuart123** [comment]
So ye @mudrila I've hit a bit of a wall

**editing my old payroll (1 token over 1 week), led to the below** 
i) I think the 1 token is appearing in wei as 0.00000000000000001
ii) I can't get the 'invalid amount' to go away with any number in the amount

Screenshot 2024-07-22 at 16 28 14

**Then I tried creating a new payroll to then edit later**
i) however, I keep getting 'proposal failed' on proposal creation. The treasury/moralis is showing up okay, so I think its somehting else blocking me
Screenshot 2024-07-22 at 16 31 32


I can try creating a payroll on the old PR and then editing it here, but thought I'd leave this feedback so you can start on it

**Comment by tomstuart123** [comment]
sorry @mudrila but I mean that when I try and change the wearer, the CTA for 'Save' is greyed out

i.e. all I did was update the below
Screenshot 2024-07-22 at 18 54 46

to this below 
Screenshot 2024-07-22 at 18 52 47

maybe I'm missing something?

**Comment by nicolaus-sherrill** [comment]
Created a new issue [Tab navigation breaks proposal preview. #2170](https://github.com/decentdao/decent-interface/issues/2170) in GitHub

**Comment by nicolaus-sherrill** [comment]
screenshot

I'm curious why it is not possible to select a date in the past to start payroll. We learned that streams can be backdated and backfilled accordingly. It's not exactly my first instinct to select a date in the past, but it _is_ possible.

<details><summary>Browser metadata</summary>





```
Path:      /roles/edit?dao=sep:0xe20221A8939F9373C091262CBEF9493aca196acD
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1680 x 904 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%2Fedit%3Fdao%3Dsep%3A0xe20221A8939F9373C091262CBEF9493aca196acD&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1680x904&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit?dao=sep:0xe20221A8939F9373C091262CBEF9493aca196acD) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=669ff7ebd54a9c1abc938247&resolution=resolved)

**Comment by nicolaus-sherrill** [comment]
image
<details><summary>Browser metadata</summary>





```
Path:      /roles/edit/details?dao=sep:0xe20221A8939F9373C091262CBEF9493aca196acD&hatId=0x8f14e62d27de72d05967837bc1ef351734c57db8f877db60f6913ff74373a5ae
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1680 x 904 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%2Fedit%2Fdetails%3Fdao%3Dsep%3A0xe20221A8939F9373C091262CBEF9493aca196acD%26hatId%3D0x8f14e62d27de72d05967837bc1ef351734c57db8f877db60f6913ff74373a5ae&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1680x904&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit/details?dao=sep:0xe20221A8939F9373C091262CBEF9493aca196acD&hatId=0x8f14e62d27de72d05967837bc1ef351734c57db8f877db60f6913ff74373a5ae) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=669ffb00ce3bfd273ef0d0d6&resolution=resolved)

**Comment by nicolaus-sherrill** [comment]
Screenshot 2024-07-23 at 1 49 12 PM

z-index issue with tooltip and `Select Asset` dropdown

**Comment by nicolaus-sherrill** [comment]
https://github.com/user-attachments/assets/2a9d8628-539f-40cd-9a52-00565cf05c2c

Two things here:
1. Dropdown is expanding from the right hand corner. Looks really odd. Let's have that drop directly down below. A slide down effect.

2. Dropdown selector styling looks off. It should have the same raised dimensional effect, but looks recessed instead.

**Comment by nicolaus-sherrill** [comment]
Screenshot 2024-07-23 at 2 05 37 PM

Abbreviation for "week" is "wk"

**Comment by nicolaus-sherrill** [comment]
Screenshot 2024-07-23 at 2 08 13 PM

Desktop side drawer should be using same overlay as other drawers.

Styling on drop shadow should only be 1px spread with no blur. Needs dimensional styling.

**Comment by nicolaus-sherrill** [comment]
Screenshot 2024-07-23 at 2 10 41 PM

Alignment of property labels and text input fields is off. Can we center up this proposal editor?

Extra padding on the card container for the affected roles.

**Comment by xraystyle1980** [comment]
screenshot

I created a these 2 roles at the same time/same proposal however none of the payment + vesting was applied. (the admin role should have payroll + vesting and the designer role should have payroll)


<details><summary>Browser metadata</summary>





```
Path:      /roles?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1708 x 1027 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%3Fdao%3Dsep%3A0x316ce0Bd801bC39A19E531d55762847BCad1201c&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1708x1027&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a0077ed54a9c2b2d9383a5&resolution=resolved)

**Comment by xraystyle1980** [comment]
screenshot

When I enter edit mode, the table columns shift to the left. I assume this is to make room for the status badge. But I wonder if we could handle this change a little more elegantly? i.e., less of a dramatic shift

<details><summary>Browser metadata</summary>





```
Path:      /roles/edit?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1465 x 1027 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%2Fedit%3Fdao%3Dsep%3A0x316ce0Bd801bC39A19E531d55762847BCad1201c&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1465x1027&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a00883036db235850a725b&resolution=resolved)

**Comment by xraystyle1980** [comment]
screenshot


<details><summary>Browser metadata</summary>





```
Path:      /roles/edit/details?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c&hatId=0x0000019400010001000000000000000000000000000000000000000000000000
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1708 x 1027 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%2Fedit%2Fdetails%3Fdao%3Dsep%3A0x316ce0Bd801bC39A19E531d55762847BCad1201c%26hatId%3D0x0000019400010001000000000000000000000000000000000000000000000000&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1708x1027&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit/details?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c&hatId=0x0000019400010001000000000000000000000000000000000000000000000000) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a0094e5b0d382e35d21ae0&resolution=resolved)

**Comment by nicolaus-sherrill** [comment]
Screenshot 2024-07-23 at 3 01 48 PM
Assets showing completely blank on Sepolia Safe.

Treasury holds Safe voting token, Sepolia ETH, Sepolia USDC

**Comment by nicolaus-sherrill** [comment]
> screenshot
> 
> I created a these 2 roles at the same time/same proposal however none of the payment + vesting was applied. (the admin role should have payroll + vesting and the designer role should have payroll)
> 
> Browser metadata
> [Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a0077ed54a9c2b2d9383a5&resolution=resolved)

I experienced this same issue.

**Comment by nicolaus-sherrill** [comment]
> @mudrila - great question. I don't think enough thought has been put into this.
> 
> **Before I choose,**
> 
> 1. I think regardless of the option, both options should have more user feedback during creation to make sure its clear.
> 
> I.e. if we want the user to input '1000 tokens' + '10 months', we should dynamically generate some copy saying 'That'll be 100 per month'. OR if we go the other path, if we want the user to input '100 tokens' over '10 months', we should generate some copy saying 'That'll remove 1000 tokens from the treasury'
> 
> 2. I think the copy here and here both need clarifying to make it easier to understand
>    Screenshot 2024-07-22 at 14 44 21
>    Screenshot 2024-07-22 at 14 47 18
> 
> **Now my take on choice**
> 
> **re; Vesting** - this should 100% be the 'Total Amount'. Very few people think about vested tokens in a bottom up way
> 
> **re; Payroll -** this could be either way... but I lean to the bottom up way due to i) I assume multiplication is easier than division (don't have to worry about decimals) and ii) when contractors quote their pay, they quote how much $x every y month. HOWEVER, this path relies on the dynamic text above (i.e. they could get confused if getting an 'invalid' message due to the total being bigger than what is in the treasury)
> 
> Note - @mudrila - if it's easier to make both Total Amount bottom down as a start, we can make this a seperate PR
> 
> @nicolaus-sherrill what's your take?

Getting around to this now.

I like what you said about contractors thinking of their payments in terms of `$x every y`. I was thinking the same. With that said, I'm leaning toward the logic supporting not the _total_ amount of pay over the course of the stream, but rather the amount per streaming increment.

So encouraging the user to input $5,000 /mo for 6 months. Maybe "Total Amount" label gets changed to simply "Amount", and somewhere in the UI there is an indication of the actual _total_ amount over the life of the stream.

**Review by nicolaus-sherrill — CHANGES_REQUESTED** [review]
A lot of visual issues to be polished up, and I could not get a payroll to be created.
Buuuut other than that, the new UI is looking great y'all, and it's very helpful to see these flows in action to get a better feel for them for UX.

**Comment by xraystyle1980** [comment]
screenshot


Please add footer component below submit proposal button
<details><summary>Browser metadata</summary>





```
Path:      /roles/edit/summary?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1465 x 1027 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%2Fedit%2Fsummary%3Fdao%3Dsep%3A0x316ce0Bd801bC39A19E531d55762847BCad1201c&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1465x1027&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit/summary?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a0120b209146382d4b6d9b&resolution=resolved)

**Comment by xraystyle1980** [comment]
recording

Getting Toast message with "Unable to encode proposal data" (see recording)

<details><summary>Browser metadata</summary>





```
Path:      /roles/edit/summary?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1708 x 1027 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Froles%2Fedit%2Fsummary%3Fdao%3Dsep%3A0x316ce0Bd801bC39A19E531d55762847BCad1201c&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1708x1027&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit/summary?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a01352ad6e393870041784&resolution=resolved)

**Comment by xraystyle1980** [comment]
recording

At the end of the recording, after the "Vote successful" toast goes away, the voting controls will briefly flash. Is the intended behavior to disappear completely?
<details><summary>Browser metadata</summary>





```
Path:      /proposals/2?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c
Browser:   Chrome 126.0.0.0 on Mac OS 10.15.7
Viewport:  1708 x 1027 @2x
Language:  en-US
Cookies:   Enabled
```

[Open in BrowserStack](https://www.browserstack.com/user/try-live?url=https%3A%2F%2Fdeploy-preview-2165.app.dev.decentdao.org%2Fproposals%2F2%3Fdao%3Dsep%3A0x316ce0Bd801bC39A19E531d55762847BCad1201c&os=OS%20X&os_version=Catalina&browser=Chrome&browser_version=126.0&resolution=1708x1027&speed=1&start=true&ref=netlify-source)



</details>

[Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/proposals/2?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a01c0e4c78b342bd2dfed2&resolution=resolved)

**Comment by adamgall** [comment]
> recording
> 
> Getting Toast message with "Unable to encode proposal data" (see recording)
> 
> Browser metadata
> [Open Deploy Preview](https://deploy-preview-2165.app.dev.decentdao.org/roles/edit/summary?dao=sep:0x316ce0Bd801bC39A19E531d55762847BCad1201c) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=669fde846e74d00008449ecc&commentID=66a01352ad6e393870041784&resolution=resolved)

@xraystyle1980 is this an Azorius DAO, and did you delegate tokens to yourself before attempting to create that proposal?

**Comment by Da-Colon** [comment]
> @tomstuart123 @nicolaus-sherrill @adamgall
> 
> One caveat on displaying payroll that I noticed - on the code level I'm using `Total Amount` that user provides as a whole amount of the stream. Which means if you were putting `12000` there and set frequency to "12 months" - this will display `1000 / mo`. I've raised this before on Figma, and just wanted to confirm - this is how we wanna this to be displayed. Currently it creates misleading UI/UX during payroll creation, since it will display `12000 / mo` while in fact that will be `1000 / mo`.
> 
> My question here :
> 
> * Should we adjust the logic so that amount that users puts is then properly divided per number of payroll tranches, aka user puts 12000, we dividing that to figure out monthly/bi-weekly/weekly tranche amount, and then displaying `1000 / mo`
> * Or should we use that "Total Amount" field as "tranche amount" and on the code level "reverse logic" - figure out total amount to be streamed based on "tranche amount" multiplied by number of pay cycles

I think what you are saying is the current expectation. Just design transfer the same number instead of 5000/12. @nicolaus-sherrill. Can you confirm this. I was also thinking that @Delmslaugh could take a look at this workflow and see if it makes sense from a admin prospective.

I guess the question is..
- Is the `Total Amount` field the Total Amount that will be divided by the `frequency` option the user chooses. 
or
- Should it be `amount` that will be claimable at each period. 

I guess this UX might change up a little per today's discussion, just wanted to note my thoughts on current setup adn comment

File: src/hooks/DAO/loaders/useHatsTree.ts [hunk 1]
```diff
@@ -132,12 +139,143 @@ const useHatsTree = () => {
...
+              let vesting: SablierVesting | undefined;
+
+              if (!streamQueryResult.data.streams.length) {
+                return hat;
+              }
+
+              const activeStreams = streamQueryResult.data.streams.filter(
+                stream =>
+                  parseInt(stream.endTime) <= Date.now() / 1000 ||
+                  BigInt(stream.withdrawnAmount) !== BigInt(stream.intactAmount),
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/DAO/loaders/useHatsTree.ts:174]: I think this is correct. I read this as

`in active period` or `there is money to be claimed`

File: src/hooks/streams/useSablierStream.ts [hunk 2]
```diff
@@ -199,6 +210,40 @@ export default function useCreateSablierStream() {
...
+      functionName: 'withdrawMax',
+      args: [BigInt(stream.streamId), to],
+    });
+
+    return { calldata: flushCalldata, targetAddress: stream.contractAddress };
+  }, []);
+
+  const prepareCancelStreamTx = useCallback((stream: BaseSablierStream) => {
+    if (!stream.streamId || !stream.contractAddress) {
+      throw new Error('Can not flush stream without streamId or contract address');
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/hooks/streams/useSablierStream.ts:240]: yessir

**Comment by Da-Colon** [comment]
> sorry @mudrila but I mean that when I try and change the wearer, the CTA for 'Save' is greyed out
> 
> i.e. all I did was update the below Screenshot 2024-07-22 at 18 54 46
> 
> to this below Screenshot 2024-07-22 at 18 52 47
> 
> maybe I'm missing something?

Ah I'm able to reproduce. It does have something to do with edit side of things. this will probably change after today's discussion

**Pull request merged**